pkg/codegen/importer/hcl2.go

```diff
@@ -0,0 +1,467 @@
+		return 3
+	case schema.StringType:
+		return 4
+	case schema.ArchiveType:
+		return 5
+	case schema.AssetType:
+		return 6
+	case schema.JSONType:
+		return 7
+	case schema.AnyType:
+		return 13
+	default:
+		switch t.(type) {
+		case *schema.TokenType:
+			return 8
+		case *schema.ArrayType:
+			return 9
+		case *schema.MapType:
+			return 10
+		case *schema.ObjectType:
+			return 11
+		case *schema.UnionType:
+			return 12
+		default:
+			return int(math.MaxInt32)
+		}
+	}
+}
+
+// simplerType returns true if T is simpler than U.
+//
+// The first-order ranking is:
+//
+//   bool < int < number < string < archive < asset < json < token < array < map < object < union < any
+//
+// Additional rules apply to composite types of the same kind:
+// - array(T) is simpler than array(U) if T is simpler than U
+// - map(T) is simpler than map(U) if T is simpler than U
+// - object({ ... }) is simpler than object({ ... }) if the former has a greater number of required properties that are
+//   simpler than the latter's required properties
+// - union(...) is simpler than union(...) if the former's simplest element type is simpler than the latter's simplest
+//   element type
+func simplerType(t, u schema.Type) bool {
+	tRank, uRank := typeRank(t), typeRank(u)
+	if tRank < uRank {
+		return true
+	} else if tRank > uRank {
+		return false
+	}
+
+	switch t := t.(type) {
+	case *schema.TokenType:
+		u := u.(*schema.TokenType)
+		if t.UnderlyingType != nil && u.UnderlyingType != nil {
+			return simplerType(t.UnderlyingType, u.UnderlyingType)
+		}
+		return false
+	case *schema.ArrayType:
+		return simplerType(t.ElementType, u.(*schema.ArrayType).ElementType)
+	case *schema.MapType:
+		return simplerType(t.ElementType, u.(*schema.ArrayType).ElementType)
+	case *schema.ObjectType:
+		// Count how many of T's required properties are simpler than U's required properties and vice versa.
+		uu := u.(*schema.ObjectType)
+		tscore, nt, uscore := 0, 0, 0
+		for _, p := range t.Properties {
+			if p.IsRequired {
+				nt++
+				for _, q := range uu.Properties {
+					if q.IsRequired {
+						if simplerType(p.Type, q.Type) {
+							tscore++
+						}
+						if simplerType(q.Type, p.Type) {
+							uscore++
+						}
+					}
+				}
+			}
+		}
+
+		// If the number of T's required properties that are simpler that U's required properties exceeds the number
+		// of U's required properties that are simpler than T's required properties, T is simpler.
+		if tscore > uscore {
+			return true
+		}
+		if tscore < uscore {
+			return false
+		}
+
+		// If the above counts are equal, T is simpler if it has fewer required properties.
+		nu := 0
+		for _, q := range uu.Properties {
+			if q.IsRequired {
+				nu++
+			}
+		}
+
+		return nt < nu
+	case *schema.UnionType:
+		// Pick whichever has the simplest element type.
+		var simplestElementType schema.Type
+		for _, t := range t.ElementTypes {
+			if simplestElementType == nil || simplerType(t, simplestElementType) {
```

Comment:
It doesn't look like `simplestElementType` has been assigned before this check the first time. Does `simplestElementType == nil` still make sense?

Reply:
Yes: the zero value for `schema.Type` is `nil`.